lib/valueflow.cpp

```diff
@@ -1951,7 +1951,7 @@ static void valueFlowGlobalStaticVar(TokenList *tokenList, const Settings *setti
 ValuePtr<Analyzer> makeAnalyzer(const Token* exprTok, ValueFlow::Value value, const TokenList* tokenlist);
 ValuePtr<Analyzer> makeReverseAnalyzer(const Token* exprTok, const ValueFlow::Value& value, const TokenList* tokenlist);
 
-static Analyzer::Result valueFlowForward(Token* startToken,
+static Analyzer::Result valueFlowForward(const Token* startToken,
```

Comment:
This shouldn't be changed to `const` as the function needs a non-const token. This will also hide the areas where we need to fix `const_cast`s that we are doing.

Reply:
The previous code was doing much more hiding since you had to do it all over the place and with the pointer never being const there's no indication why they even have to be mutable. This clears this up greatly. Only `setTokenValues()` requires it and there's no need to make those pointers non-const along the whole way.

It's the same with moving mutable objects around although there's just two occurrences deeply buried somewhere in the code which actually needs it.

Reply:
> The previous code was doing much more hiding since you had to do it all over the place

It wasnt hiding, it was making the non-const explicit, and showing places in the code where we should be using a non-const token.

> with the pointer never being const there's no indication why they even have to be mutable.

It has to be mutable as it modifies the `Token` class. This also allow compile-checking to ensure we dont modify the tokens in contexts where we shouldn't. For example, we dont want to call `valueFlowForward` or `setTokenValues` in the `analyze` function which is why it is not `const`(whereas `update` is non-const so such operations are permitted). However, this change takes away this type-safety. 

> Only setTokenValues() requires it 

That really is an implementation detail. We could add other functions in the future to modify the Token class as well.

> there's no need to make those pointers non-const along the whole way.

I disagree. We should have the Tokens non-const from the beginning of traversal so we can get the extra safety that comes with `const`. We are wanting to modify the `Token` classes from the beginning so why wouldn't we use a non-const `Token`?

> although there's just two occurrences deeply buried somewhere in the code which actually needs it.

Its not about which functions we are calling to mutate the `Token`. Its about where we want to permit such mutations and where we want to prohibit it(especially since in some places we have several layers of abstractions so it might not always be obvious).